resolwe/test/testcases/process.py

```diff
@@ -729,6 +750,127 @@ def assertFilesExist(self, obj, field_path):  # pylint: disable=invalid-name
+                test_dirs.add(os.path.relpath(os.path.join(root, test_dir), dir_path))
+            for test_file in files:
+                test_files.add(os.path.relpath(os.path.join(root, test_file), dir_path))
+
+        def get_dirs(indict, root=''):
+            """Generate directory names from dict."""
+            for key, value in indict.items():
+                if isinstance(value, dict):
+                    yield os.path.join(root, key)
+                    yield from get_dirs(value, os.path.join(root, key))
+                elif value is not None:
+                    self.fail(msg='Directory structure specification is incorrect')
+
+        def get_files(indict, root=''):
+            """Generate file names from dict."""
+            for key, value in indict.items():
+                if value is None:
+                    yield os.path.join(root, key)
+                elif isinstance(value, dict) and value:
+                    yield from get_files(value, os.path.join(root, key))
+                elif not isinstance(value, dict):
+                    self.fail(msg='Directory structure specification is incorrect.')
+
+        correct_files = {file for file in get_files(dir_struct)}
+        correct_dirs = {correct_dir for correct_dir in get_dirs(dir_struct)}
+
+        if exact and (test_dirs != correct_dirs or test_files != correct_files):
+            self.fail(msg='Directory structure mismatch (exact check).')
+        if not exact and (correct_dirs - test_dirs or correct_files - test_files):
+            self.fail(msg='Directory structure mismatch (partial structure check).')
+
+    def assertDirStructure(self, obj, field_path, dir_struct, exact=True):  # pylint: disable=invalid-name
+        """Assert correct tree structure in output field of given object.
+
+        Only names of directories and files are asserted. Content of files is
+        not compared.
+
+        :param obj: object that includes the directory to compare
+        :type obj: ~resolwe.flow.models.Data
+
+        :param str dir_path: path to the directory to compare
+
+        :param dict dir_struct: correct tree structure of the directory.
+            Dictionary keys are directory and file names with the correct nested
+            structure. Dictionary value associated with each directory is a new
+            dictionary which lists the content of the directory. Dictionary
+            value associated with each file name is ``None``
+
+        :param bool exact: if ``True`` tested directory structure must exactly
+            match `dir_struct`. If ``False`` `dir_struct` must be a partial
+            structure of the directory to compare
+
+        """
+        self.assertDirExists(obj, field_path)
+        field = dict_dot(obj.output, field_path)
+        dir_path = obj.location.get_path(filename=field['dir'])
+        self._assert_dir_structure(dir_path, dir_struct, exact)
+
+    def _assert_dir(self, dir_path, fn_correct, fail_on_funny=True):
```

Comment:
Is `fail_on_funny` still needed?

Reply:
I would leave it in in case we encounter something nasty, but we don't have a specific use-case at the moment.